File: src/routes/bot.js
const express = require('express');
const db = require('../config/database');
const { authenticateAgent } = require('../middleware/auth');

const router = express.Router();

// GET /api/bot/config
router.get('/bot/config', authenticateAgent, async (req, res) => {
  try {
    const r = await db.query('SELECT instrucciones, activo_global, contexto FROM bot_config WHERE id = 1');
    const row = r.rows[0] || { instrucciones: '', activo_global: true, contexto: {} };
    res.json({ ...row, contexto: row.contexto || {} });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

// PUT /api/bot/config
router.put('/bot/config', authenticateAgent, async (req, res) => {
  try {
    const { instrucciones, activo_global, contexto } = req.body;

    const sets = ['updated_at = NOW()'];
    const params = [];
    let i = 1;

    if (instrucciones !== undefined) { sets.push(`instrucciones = $${i++}`); params.push(instrucciones); }
    if (activo_global !== undefined) { sets.push(`activo_global = $${i++}`); params.push(Boolean(activo_global)); }
    if (contexto     !== undefined) { sets.push(`contexto = $${i++}`);      params.push(JSON.stringify(contexto)); }

    await db.query(`UPDATE bot_config SET ${sets.join(', ')} WHERE id = 1`, params);
    res.json({ ok: true });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

// GET /api/bot/faq
router.get('/bot/faq', authenticateAgent, async (req, res) => {
  try {
    const r = await db.query('SELECT * FROM bot_faq WHERE activo = true ORDER BY orden, created_at');
    res.json(r.rows);
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

// POST /api/bot/faq
router.post('/bot/faq', authenticateAgent, async (req, res) => {
  try {
    const { pregunta, respuesta } = req.body;
    if (!pregunta || !respuesta) return res.status(400).json({ error: 'Faltan pregunta o respuesta' });
    const r = await db.query(
      'INSERT INTO bot_faq (pregunta, respuesta) VALUES ($1, $2) RETURNING *',
      [pregunta, respuesta]
    );
    res.json(r.rows[0]);
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

// DELETE /api/bot/faq/:id
router.delete('/bot/faq/:id', authenticateAgent, async (req, res) => {
  try {
    await db.query('UPDATE bot_faq SET activo = false WHERE id = $1', [req.params.id]);
    res.json({ ok: true });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

// PATCH /api/conversaciones/:id/bot  — toggle bot por conversación
router.patch('/conversaciones/:id/bot', authenticateAgent, async (req, res) => {
  try {
    const { bot_activo } = req.body;
    await db.query('UPDATE conversaciones SET bot_activo = $1 WHERE id = $2', [bot_activo, req.params.id]);
    res.json({ ok: true, bot_activo });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

module.exports = router;

File: app/routes/bot.py
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from typing import Optional
from app.middleware.auth import get_agente
from app.crud import bot as crud

router = APIRouter()


class BotConfigBody(BaseModel):
    instrucciones: Optional[str] = None
    activo_global: Optional[bool] = None
    contexto: Optional[dict] = None


class FaqBody(BaseModel):
    pregunta: str
    respuesta: str


class BotToggleBody(BaseModel):
    bot_activo: bool


class ExcluidoBody(BaseModel):
    numero: str
    motivo: Optional[str] = None


class MensajePrueba(BaseModel):
    role: str
    content: str


class PruebaBody(BaseModel):
    mensajes: list[MensajePrueba]


@router.get("/api/bot/config")
def get_config(agente=Depends(get_agente)):
    return crud.get_config()


@router.put("/api/bot/config")
def update_config(body: BotConfigBody, agente=Depends(get_agente)):
    crud.update_config(body.model_dump(exclude_none=True))
    return {"ok": True}


@router.delete("/api/bot/config")
def reset_config(agente=Depends(get_agente)):
    crud.reset_config()
    return {"ok": True}


@router.get("/api/bot/faq")
def list_faq(agente=Depends(get_agente)):
    return crud.list_faq()


@router.post("/api/bot/faq")
def create_faq(body: FaqBody, agente=Depends(get_agente)):
    if not body.pregunta or not body.respuesta:
        raise HTTPException(400, detail="Faltan pregunta o respuesta")
    return crud.create_faq(body.pregunta, body.respuesta)


@router.delete("/api/bot/faq/{faq_id}")
def delete_faq(faq_id: str, agente=Depends(get_agente)):
    crud.delete_faq(faq_id)
    return {"ok": True}


@router.patch("/api/conversaciones/{conv_id}/bot")
def toggle_bot(conv_id: str, body: BotToggleBody, agente=Depends(get_agente)):
    from app.crud.conversaciones import toggle_bot as crud_toggle
    crud_toggle(conv_id, body.bot_activo)
    return {"ok": True, "bot_activo": body.bot_activo}


# ── Números excluidos ──

@router.get("/api/bot/excluidos")
def list_excluidos(agente=Depends(get_agente)):
    return crud.list_excluidos()


@router.post("/api/bot/excluidos", status_code=201)
def add_excluido(body: ExcluidoBody, agente=Depends(get_agente)):
    if not body.numero.strip():
        raise HTTPException(400, detail="El número es requerido")
    return crud.add_excluido(body.numero, body.motivo)


@router.delete("/api/bot/excluidos/{excluido_id}")
def delete_excluido(excluido_id: str, agente=Depends(get_agente)):
    if not crud.delete_excluido(excluido_id):
        raise HTTPException(404, detail="No encontrado")
    return {"ok": True}


# ── Modo prueba: chatear con el bot sin tocar BD ni WhatsApp ──

@router.post("/api/bot/probar")
async def probar_bot(body: PruebaBody, agente=Depends(get_agente)):
    from fastapi.concurrency import run_in_threadpool
    from app.services.ai.index import responder_prueba
    historial = [{"role": m.role, "content": m.content} for m in body.mensajes]
    respuesta = await run_in_threadpool(responder_prueba, historial)
    return {"respuesta": respuesta}
